// codesandboxer/dist/fetchFiles/finaliseCSB.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

exports.default = function (_ref, providedFiles, passedDeps, name) {
  var files = _ref.files,
      deps = _ref.deps;

  var dependencies = (0, _extends3.default)({}, deps, passedDeps);

  ensureReact(dependencies);

  var finalFiles = (0, _extends3.default)({}, files, {
    'package.json': {
      content: (0, _constants.newpkgJSON)(dependencies, name)
    }
  }, providedFiles);
  var parameters = (0, _define.getParameters)({
    files: finalFiles,
    template: 'create-react-app'
  });
  return {
    files: finalFiles,
    dependencies: dependencies,
    parameters: parameters
  };
};

var _define = require('codesandbox/lib/api/define');

var _constants = require('../constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ensureReact = function ensureReact(deps) {
  if (!deps.react && !deps['react-dom']) {
    deps.react = 'latest';
    deps['react-dom'] = 'latest';
  } else if (!deps.react) {
    deps.react = deps['react-dom'];
  } else if (!deps['react-dom']) {
    deps['react-dom'] = deps.react;
  }
};
