// codesandboxer/dist/finaliseCSB/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

exports.default = function (_ref, config) {
  var files = _ref.files,
      deps = _ref.deps;

  if (!config) config = {};
  var _config = config,
      extraFiles = _config.extraFiles,
      extraDependencies = _config.extraDependencies,
      name = _config.name,
      _config$template = _config.template,
      template = _config$template === undefined ? 'create-react-app' : _config$template,
      main = _config.main;

  var dependencies = (0, _extends3.default)({}, deps, extraDependencies);

  main = !main && template === 'create-react-app-typescript' ? 'index.tsx' : 'index.js';

  ensureReact(dependencies);

  var finalFiles = (0, _extends3.default)({}, files, {
    'package.json': {
      content: (0, _constants.newpkgJSON)(dependencies, name, main)
    },
    'sandbox.config.json': {
      content: (0, _stringify2.default)({
        template: template
      })
    }
  }, extraFiles);
  var parameters = (0, _getParameters2.default)({
    files: finalFiles
  });
  return {
    files: finalFiles,
    dependencies: dependencies,
    parameters: parameters
  };
};

var _getParameters = require('./getParameters');

var _getParameters2 = _interopRequireDefault(_getParameters);

var _constants = require('../constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ensureReact = function ensureReact(deps) {
  if (!deps.react && !deps['react-dom']) {
    deps.react = 'latest';
    deps['react-dom'] = 'latest';
  } else if (!deps.react) {
    deps.react = deps['react-dom'];
  } else if (!deps['react-dom']) {
    deps['react-dom'] = deps.react;
  }
};
