// codesandboxer/dist/constants.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getSandboxUrl = exports.codesandboxURLJSON = exports.codesandboxURL = exports.newpkgJSON = exports.baseFilesTS = exports.baseFiles = undefined;

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var baseFiles = exports.baseFiles = {
  'index.html': {
    content: '<div id="root"></div>'
  },
  'index.js': {
    content: '/**\n  This CodeSandbox has been automatically generated using\n  `codesandboxer`. If you\'re curious how that happened, you can\n  check out our docs here: https://github.com/noviny/codesandboxer\n\n  If you experience any struggles with this sandbox, please raise an issue\n  on github. :)\n*/\nimport React from \'react\';\nimport ReactDOM from \'react-dom\';\nimport App from \'./example\';\n\nReactDOM.render(\n<App />,\ndocument.getElementById(\'root\')\n);'
  }
};
var baseFilesTS = exports.baseFilesTS = {
  'index.html': {
    content: '<div id="root"></div>'
  },
  'index.tsx': {
    content: '/**\n  This CodeSandbox has been automatically generated using\n  `codesandboxer`. If you\'re curious how that happened, you can\n  check out our docs here: https://github.com/noviny/codesandboxer\n\n  If you experience any struggles with this sandbox, please raise an issue\n  on github. :)\n*/\nimport * as React from \'react\';\nimport * as ReactDOM from \'react-dom\';\nimport App from \'./example\';\n\nReactDOM.render(\n<App />,\ndocument.getElementById(\'root\')\n);'
  }
};

var newpkgJSON = exports.newpkgJSON = function newpkgJSON(dependencies) {
  var name = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'codesandboxer-example';
  var main = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'index.js';
  return '{\n  "name": "' + name + '",\n  "version": "0.0.0",\n  "description": "A simple example deployed using react-codesandboxer",\n  "main": "' + main + '",\n  "dependencies": {\n    ' + (0, _keys2.default)(dependencies).map(function (k) {
    return '"' + k + '": "' + dependencies[k] + '"';
  }).join(',\n    ') + '\n  }\n}';
};

var codesandboxURL = exports.codesandboxURL = 'https://codesandbox.io/api/v1/sandboxes/define?query=module=/example.js';
var codesandboxURLJSON = exports.codesandboxURLJSON = 'https://codesandbox.io/api/v1/sandboxes/define?json=1';

var getSandboxUrl = exports.getSandboxUrl = function getSandboxUrl(id) {
  var type = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 's';
  return 'https://codesandbox.io/' + type + '/' + id + '?module=/example';
};
